### DISCUSSION TITLE
Get pull request associated with a commit

### DISCUSSION THREAD

**Discussion by TimonVS** [community/community#24679] [Apps, API and Webhooks]
<p>I was wondering if it’s possible to get a pull request associated with a commit. With a merge-commit or squash-merging I can infer the associated pull request from the commit message, but in the case of rebase-merging the commits are directly appended to the master branch which makes this impossible. </p>
<p>It seems like GitHub has a way of associating commits that have been merged using rebase-merging as can be seen in the following screenshot (all appended commits show this):</p>
<p><img src="https://user-images.githubusercontent.com/876666/181086016-b5d1ca7a-4d6b-406a-93da-037ed3ccc00e.png" alt="pr-number.png" data-base62-sha1="9p78SwJnQk4WTKZHFJ4aWiT3Rga" width="372" height="156"></p>
<p>I’ve been digging through the V4 and V3 API but I couldn’t find a way to associate commits with a PR. Is this somehow possible?</p>
<p>For more context: I’m currently working on supporting rebase-merging in Release Drafter (<a href="https://github.com/toolmantim/release-drafter/issues/106" rel="nofollow noopener">https://github.com/toolmantim/release-drafter/issues/106</a>).</p>

**Comment by AndreaGriffiths11** [comment]
<p>Hey <a class="mention" href="/timonvs">@timonvs</a>, Thanks for being here! This is a great question,  you can do this using the search UI but I don’t think there is a way to do it programmatically via the API. I’ll dig a little dipper and post any updates here. Hopefully another user might have more insight into this question.</p>

**Comment by TimonVS** [comment]
<p>Thanks <a class="mention" href="/andreagriffiths11">@andreagriffiths11</a>. It’s indeed possible with the search API (<a href="https://help.github.com/articles/searching-issues-and-pull-requests/#search-by-commit-sha" rel="nofollow noopener">https://help.github.com/articles/searching-issues-and-pull-requests/#search-by-commit-sha</a>). For my use case however that would not work very well because I need be able to identify PRs from a bunch of commits. Doing a request for every commit would be very ineffecient and I’d quickly hit the API limits.</p>
<p>Ideally I’d be able to do the following GraphQL query:</p>
<p>  </p>
<pre><code>{
  repository(owner: "awkward", name: "sketch-frontend") {
    ref(qualifiedName: "master") {
      target {
        ... on Commit {
          id
          associatedPullRequest {
            title
            number
          }
        }
      }
    }
  }
}
</code></pre>

**Comment by AndreaGriffiths11** [comment]
<p><a class="mention" href="/timonvs">@timonvs</a> thanks for clarifying, I believe that schema is not supported yet, but has been requested. </p>

**Comment by caleb15** [comment]
<p>I figured out how to do this using <a href="https://github3py.readthedocs.io/en/master/" rel="nofollow noopener">github3.py</a>. It assumes you have cloned the repo locally.</p>
<pre><code>import github3
import subprocess
import os

os.chdir('C:\\yourRepo')

github_repo_owner = '15five'
github_repo_name = 'yourRepo'
sha = '1425ca3' # replace with desired sha

# thanks to https://stackoverflow.com/questions/17818167
out = subprocess.check_output(['git', 'describe', '--all', '--contains', sha])
text = out.decode('utf-8')
branch = text.replace('remotes/origin/','').split('~')[0]

gh = github3.GitHub(token='CENSORED') # replace token
repo = gh.repository(github_repo_owner, github_repo_name)
pulls = repo.pull_requests(state='closed', sort='updated', head=github_repo_owner+':'+branch, number=1)
your_pr = next(pulls)
print(your_pr.title)
</code></pre>

**Comment by TimonVS** [comment, marked as answer]
<p>I was just randomly looking at the GraphQL API changelog and I saw that the <code>associatedPullRequests</code> field has recently been added (<a href="https://developer.github.com/v4/changelog/2019-03-08-schema-changes/" rel="nofollow noopener">https://developer.github.com/v4/changelog/2019-03-08-schema-changes/</a>) which resolves this issue. Thanks a lot <a class="mention" href="/andreagriffiths11">@andreagriffiths11</a> :pray:</p>

**Comment by TimonVS** [comment]
<p>Here’s an example query which demonstrates how to fetch associated pull requests of the five latest commits on the master branch:</p>
<pre><code>{
  repository(name: "react", owner: "facebook") {
    ref(qualifiedName: "master") {
      target {
        ... on Commit {
          id
          history(first: 5) {
            nodes {
              id
              associatedPullRequests(first: 10) {
                edges {
                  node {
                    title
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
</code></pre>

**Reply by windchasm** [reply]
Do you know how to fetch the associated pull requests for a *specific* commit hash that has _not been merged yet_? For example, for the following repository (https://github.com/SimpleServer/SimpleServer), I want to fetch all the associated pull requests in which the commit with hash `e15978c336e07c3209cb957b5ded911161db7cf9` appears.

So far, I've tried something like this:
```
{
  repository(owner: "SimpleServer", name: "SimpleServer") {
    ref(qualifiedName: "master") {
      target {
        ... on Commit {
          url
          associatedPullRequests(first: 50) {
            edges {
              node {
                number
                title
                closedAt
                mergedAt
                createdAt
              }
            }
          }
        }
      }
    }
  }
}
```

But it only returns results for the latest commit (which is empty, since it appears in no pull requests)...

The reason why I am specifying for a commit that hasn't been merged yet is because using the `associatedPullRequests` field for the `Commit` object only returns results for _merged_ pull requests (as I understand it from the documentation: https://docs.github.com/en/graphql/reference/objects#commit). However, using the `associatedPullRequests` field for the `Ref` object seems to not be limited to only merged pull requests.

**Comment by caleb15** [comment]
<p>Here is another example graphql query that takes a specific sha and outputs the associated PR’s:</p>
<pre><code>query associatedPRs($sha: String, $repo: String!, $owner: String!){
  repository(name: $repo, owner: $owner) {
    commit: object(expression: $sha) {
      ... on Commit {
        associatedPullRequests(first:5){
          edges{
            node{
              title
              number
              body
            }
          }
        }
      }
    }
  }
}

{
                    "sha": 14242132153asfe33453,
                    "repo": "repo name",
                    "owner": "repo owner"
}
</code></pre>

**Comment by eltronix** [comment]
<p>Thanks! precisely what I was looking for.</p>
<p>So commits are abstracted behind GitObjects 🤭</p>

**Comment by kod-man** [comment]
<p><a href="https://github.community/t/orderbyfield-issue/215084" class="inline-onebox">OrderByField issue</a>  <a class="mention" href="/caleb15">@caleb15</a> <a class="mention" href="/timonvs">@timonvs</a>  can you guys help me?</p>
